### PR TITLE
Fixed #30393 -- Added validation of startapp's directory option.

### PR DESCRIPTION
**Pull request by MyungSeKyo**
https://code.djangoproject.com/ticket/30393

### CODE REVIEW

File: django/core/management/templates.py [hunk 1]
```diff
@@ -74,6 +74,8 @@ def handle(self, app_or_project, name, target=None, **options):
             except OSError as e:
                 raise CommandError(e)
         else:
+            if app_or_project == 'app':
+                self.validate_name(os.path.basename(target))
```
Comment (felixxm) [inline django/core/management/templates.py]: I'm not sure about using `validate_name()` here, it is a bit confusing, moreover messages should be different e.g.
- `'{target}' is not a valid {app} directory. Please make sure the directory is a valid identifier.`
- `'{target}' conflicts with the name of an existing Python module and cannot be used as {an} {app} directory. Please try another directory.` (this is untested).

Maybe we can add some extra flag to distinguish messages.

File: tests/admin_scripts/tests.py [hunk 2]
```diff
@@ -2114,6 +2114,22 @@ def test_invalid_name(self):
                 )
                 self.assertFalse(os.path.exists(testproject_dir))
 
+    def test_invalid_target_name(self):
+        """startapp validates that target dir name is a valid Python identifier."""
+        os.mkdir('app_dir')
+        self.addCleanup(shutil.rmtree, 'app_dir')
```
Comment (felixxm) [inline tests/admin_scripts/tests.py]: I would use
```python
with tempfile.TemporaryDirectory() as app_dir:
     for bad_target in ('invalid.name', '7invalid_name'):
          os.mkdir(os.path.join(app_dir, bad_targer))
          ...
```

**Comment by MyungSeKyo** [comment]
@felixxm Thanks to your kind review!
I tried to fix according to your review
could you review again?

File: tests/admin_scripts/tests.py [hunk 3]
```diff
@@ -2116,19 +2116,26 @@ def test_invalid_name(self):
...
-        self.addCleanup(shutil.rmtree, 'app_dir')
-
-        for bad_target in ('app_dir/invalid.name', 'app_dir/7invalid_name'):
+        for bad_target in ('invalid.dir_name', '7invalid_dir_name'):
             args = ['startapp', 'valid_name', bad_target]
+
             os.mkdir(bad_target)
 
             out, err = self.run_django_admin(args)
+            shutil.rmtree(bad_target)
```
Comment (felixxm) [inline tests/admin_scripts/tests.py]: This line and `os.mkdir(bad_target)` are unnecessary, we don't need to create directories for this test.

File: tests/admin_scripts/tests.py [hunk 4]
```diff
@@ -2116,19 +2116,26 @@ def test_invalid_name(self):
 
     def test_invalid_target_name(self):
         """startapp validates that target dir name is a valid Python identifier."""
-        os.mkdir('app_dir')
-        self.addCleanup(shutil.rmtree, 'app_dir')
-
-        for bad_target in ('app_dir/invalid.name', 'app_dir/7invalid_name'):
+        for bad_target in ('invalid.dir_name', '7invalid_dir_name'):
             args = ['startapp', 'valid_name', bad_target]
+
```
Comment (felixxm) [inline tests/admin_scripts/tests.py]: Chop blank line.

File: tests/admin_scripts/tests.py [hunk 5]
```diff
@@ -2116,19 +2116,26 @@ def test_invalid_name(self):
...
             out, err = self.run_django_admin(args)
+            shutil.rmtree(bad_target)
             self.assertOutput(
                 err,
-                "CommandError: '{}' is not a valid app name. Please make sure the "
-                "name is a valid identifier.".format(os.path.basename(bad_target))
+                "CommandError: '{}' is not a valid app directory. Please make sure the "
+                "directory is a valid identifier.".format(bad_target)
             )
+        args = ['startapp', 'valid_name', 'os']
```
Comment (felixxm) [inline tests/admin_scripts/tests.py]: I would move test for conflict with an existing Python module to a separate test e.g. `test_importable_target_name`

File: tests/admin_scripts/tests.py [hunk 6]
```diff
@@ -2114,6 +2114,29 @@ def test_invalid_name(self):
                 )
                 self.assertFalse(os.path.exists(testproject_dir))
 
+    def test_invalid_target_name(self):
+        """startapp validates that target dir name is a valid Python identifier."""
+        for bad_target in ('invalid.dir_name', '7invalid_dir_name'):
```
Comment (jdufresne) [inline tests/admin_scripts/tests.py]: As this test runs several inner tests, you can use `unittest.subTest` to provide more useful output on each iteration.

https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest

Something like:

```py
for bad_target in ['invalid.dir_name', '7invalid_dir_name']:
    with self.subTest(bad_target):
        ...
```

**Comment by MyungSeKyo** [comment]
@jdufresne @felixxm 
Thanks to your reviews :+1: 
I made `test_invalid_target_name` and used `unittest.subTest`

**Comment by felixxm** [comment]
@MyungSeKyo Thanks! I rebased from master and pushed small final edits.

**Pull request merged**